substrate/frame/assets/src/functions.rs

```diff
@@ -193,22 +193,31 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			return Frozen
 		}
 		if let Some(rest) = account.balance.checked_sub(&amount) {
-			if let Some(frozen) = T::Freezer::frozen_balance(id.clone(), who) {
-				match frozen.checked_add(&details.min_balance) {
-					Some(required) if rest < required => return Frozen,
-					None => return Overflow,
-					_ => {},
-				}
-			}
+			match (
+				T::Holder::held_balance(id.clone(), who),
+				T::Freezer::frozen_balance(id.clone(), who),
+			) {
+				(None, None) =>
+					if rest < details.min_balance {
+						if keep_alive {
+							WouldDie
+						} else {
+							ReducedToZero(rest)
+						}
+					} else {
+						Success
+					},
+				(maybe_held, maybe_frozen) => {
+					let frozen = maybe_frozen.unwrap_or_default();
+					let held = maybe_held.unwrap_or_default();
 
-			if rest < details.min_balance {
-				if keep_alive {
-					WouldDie
-				} else {
-					ReducedToZero(rest)
-				}
-			} else {
-				Success
+					let required = frozen.saturating_sub(held).max(details.min_balance);
```

Comment:
One may think that is the case, but there are two reasons against:

1. The `frozen` part of `required` being calculated here is the `frozen` amount **in** the asset account. Since the overall `frozen` amount can overlap with `held` balance (see [Holds and Freezes](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/tokens/fungible/index.html#holds-and-freezes) section), therefore, `frozen` (in asset account) is `frozen - held` where minimum `frozen` in asset account is `0` (that's why you use `saturating`).
2. Even though the `frozen` balance in the asset account is `0`, that account is still required to maintain the minimum balance requirement in order to exist. That's why, the `max(frozen, min_balance)` is done after calculating the `frozen` in the asset account, not before.

Thus, the formula presented here (which, by the way, is exactly the same as the one that uses `pallet-balances`).

Now, about the cases you suggest to cover, have in mind when talking about decreasing a balance in the asset account, we have to see whether `rest` is greater or equal than the `spendable` (this is, `balance - untouchable`, where `untouchable` is `max(frozen - held, min_balance)`).

Personally, I'd change `Frozen` with `Unspendable`, but in this case I didn't because breaking changes (and running migrations is not that nice). That said, I'm open to add a new variant, if it helps to clarity.

Reply:
> Now, about the cases you suggest to cover, have in mind when talking about decreasing a balance in the asset account, we have to see whether `rest` is greater or equal than the `spendable` (this is, `balance - untouchable`, where `untouchable` is `max(frozen - held, min_balance)`).

isn't `spendable = balance - untouchable - on_hold` and not `spendable = balance - untouchable`

Reply:
We should have a test with both held and frozen assets.

Reply:
> isn't `spendable = balance - untouchable - on_hold` and not `spendable = balance - untouchable`

Not really. As mentioned earlier, `untouchable` is defined as `max(min_balance, frozen)`, where `frozen` can overlap `on_hold` and be less than. So, if `frozen` is less than `on_hold`, it'll be completely overlapped, and thus `untouchable` would be `min_balance` (for the sake of Assets).

> We should have a test with both held and frozen assets.

Agreed! I'll add a test to show how holds and freezes work together.

Reply:
> We should have a test with both held and frozen assets.

Agreed! I'll add a test to show how holds and freezes work together.

Reply:
OK, in pallet-asset-holder `Inspect::balance` and `Inspect::total_balance` calls the pallet-asset implementation of `Inspect::balance` and `Inspect::total_balance`.
But in pallet-asset `Inspect::balance` and `Inspect::total_balance` both calls `balance` function which returns the `balance` field for the value in the storage map `Account`.

I thought `balance` in the storage map would actually mean `total_balance`.
Now that you say `balance` in the storage map means `balance` then I think the implementation of `total_balance` is wrong and the formula above makes sense.

Reply:
see https://github.com/paritytech/polkadot-sdk/pull/4530#discussion_r1708477488

Reply:
Yeah, there are some differences between `total_balance` and `reducible_balance`. The first is the actual _asset account `balance`_. The second is the maximum `amount` that can be decreased (or `reduced`) from the `balance` with a `Success` withdrawal consequence (that is, the requirements to keep an account existing are met).

Reply:
> We should have a test with both held and frozen assets.

See [this test](https://github.com/pandres95/polkadot-sdk/blob/f2a9e96297f61580f22f28be4c990a99f51f0776/substrate/frame/assets/src/tests.rs#L1379)

Reply:
Great, thanks for your patience and explanations, formula makes sense.

Reply:
The assumption that `account.balance` does not contain `held` balance is not set as a contact/documentation anywhere on this pallet level.